### PR TITLE
new large_enum_variants lint

### PR DESCRIPTION
**Pull request by gnieto**
Added lint to detect large variants and suggesting to box them

Solves https://github.com/Manishearth/rust-clippy/issues/1127

### CODE REVIEW

File: clippy_lints/src/utils/conf.rs [hunk 1]
```diff
@@ -162,6 +162,8 @@ define_Conf! {
     ("too-large-for-stack", too_large_for_stack, 200 => u64),
     /// Lint: ENUM_VARIANT_NAMES. The minimum number of enum variants for the lints about variant names to trigger
     ("enum-variant-name-threshold", enum_variant_name_threshold, 3 => u64),
+    /// Lint: ENUM_LARGE_VARIANT. The maximum size of a emum's variant to avoid box suggestion
+    ("enum-variant-size-threshold", enum_variant_size_threshold, 200 => u64),
```
Comment (gnieto) [inline clippy_lints/src/utils/conf.rs]: Used the same value as 'too-large-for-stack'. I don't know which value use as default

Comment (mcarton) [inline clippy_lints/src/utils/conf.rs]: Is there any reason for 200? That's already quite big.

**Comment by mcarton** [comment]
r+ modulo the default value of `enum-variant-size-threshold`.

File: clippy_lints/src/enum_large_variant.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+///
+/// **Example:**
+/// ```rust
+/// enum Test {
+///    A(i32),
+///    B([i32; 8000]),
+/// }
+/// ```
+declare_lint! {
+    pub ENUM_LARGE_VARIANT,
```
Comment (mcarton) [inline clippy_lints/src/enum_large_variant.rs]: Should be named `LARGE_ENUM_VARIANT` IMO.

Comment (gnieto) [inline clippy_lints/src/enum_large_variant.rs]: Changed all references to 'enum_large_variant' to 'large_enum_variant' on 088bb8fdb2f4e8fce5a17ea4ae797d3232323b4b

File: clippy_lints/src/utils/conf.rs (hunk 1 above)
Comment (gnieto) [inline clippy_lints/src/utils/conf.rs]: So is the default value ok? Should I leave it to 200?

**Comment by oli-obk** [comment]
needs a rebase + some research into enum variant sizes in existing code bases for choosing the default value.

**Comment by gnieto** [comment]
While I was doing some research on popular Rust projects with this lint, I discovered that under some conditions, this lint will panic. Consider the following code:

```
trait SomeTrait {
    type Item;
}

enum LargeEnumGeneric<A: SomeTrait> {
    Var(A::Item),
}
```

With that code, one of the calls to rustc is panicking with:

> ../src/librustc/ty/context.rs:161: Attempted to intern `_` which contains inference types/regions in the global type context

Is obvious that the compiler is not able to know the size of 'SomeTrait::Item' because it's generic and it depends on which concrete trait implementations is given. In this scenario, the lint should ignore generic enums depending on associated types? Or the lint should search for any concretion of the enum (I don't know the technical name of this, I mean: LargeEnumGeneric<ConcreteImplementation>), and then calculate the variant sizes of this?

**Comment by mcarton** [comment]
@gnieto Good catch! IMO it's ok to bail (cleanly, not ICEing :smile:) when the size is unknown. You could manage to look at the enumeration instances, but it might be overkill.

Although, while testing this I noticed it looks like *compiletest-rs* does not catch ICE if all expected error are found before the ICE. This is *bad*, cc @Manishearth.

**Comment by Manishearth** [comment]
hmm, weird. IIRC compiletest does check the exit code.

**Comment by mcarton** [comment]
@Manishearth [*compiletest-rs*](https://github.com/laumann/compiletest-rs/blob/8bcc0454b19b628b1ba3e444e7eaff3ac0ebb0a5/src/runtest.rs#L977-L983), [*rustc*](https://github.com/rust-lang/rust/blob/2c01bb885108c436adae2006632ff6dfc0a5f2cd/src/tools/compiletest/src/runtest.rs#L979-L985)

The actual error is 
```
{all}
{other}
{errors}
{…}
{"message":"../src/librustc/ty/context.rs:161: Attempted to intern `_` which contains inference types/regions in the global type context","code":null,"level":"error: internal compiler error","spans":[],"children":[],"rendered":null}
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace
```
It looks like that function was not adapted to the JSON format. I'll do a PR tomorrow. As for the error code, it is checked but *rustc* returns *101* for any kind of error (so it should work in *run-pass*, but not *compile-fail*).

**Comment by gnieto** [comment]
It has been a while since I opened the PR , and I tried a couple of ways to avoid the internal compiler error:

- Try to get a distinct "infer context" (I was using "borrowck_fake_infer_ctxt") to avoid the ICE (without success)
- Tried to inspect recursively the TypeVariants of each field of each variant in search of associated types (TyProjection). This can lead to a infinite loop if there are types depending on each other and... well, it seems like an ugly hack.

If anyone has some idea on how I can approach the issue, I'll try to do it, but at the moment, I don't know how to continue.

**Comment by oli-obk** [comment]
> Tried to inspect recursively the TypeVariants of each field of each variant in search of associated types (TyProjection). This can lead to a infinite loop if there are types depending on each other and... well, it seems like an ugly hack.

I think that's the right approach for now. You can always use a `HashSet` to prevent recursing into types you've already seen.

**Comment by oli-obk** [comment]
@gnieto: are you planning to continue with this PR?

**Comment by gnieto** [comment]
@oli-obk I don't think that I will not have time to dedicate to this in the following weeks.
A part of that, I think that i picked a task that was too hard for me. Anyway, thanks for your help.

**Comment by oli-obk** [comment]
I just wanted to ask, to make sure I don't steal your PR that you still planned to continue. I'll finish it.

**Pull request closed without merging**